### ISSUE TITLE
Crash on ODR between instrumented and non-instrumented libraries

### ISSUE THREAD

**Issue by ramosian-glider** [google/sanitizers#398]
Originally reported on Google Code with ID 398

```
When the same global variable is defined in ASan-instrumented and in a non-instrumented
libraries, there is a 50% chance that the linker will pick the non-instrumented symbol
and __asan_register_globals will attempt to poison redzones around it.

Negative effects range from cryptic out-of-bounds reports to startup CHECK failures
(ex. because the uninstrumented variable is not 32-byte aligned).

I wonder if this can be mitigated by making the reference in asan global descriptor
point to a local symbol for the same global.

```

Reported by `eugenis@google.com` on 2015-07-14 18:09:45

**Comment by chefmax** [comment]
Well, since pr24486 was fixed (https://llvm.org/bugs/show_bug.cgi?id=24486), the patch I've posted above works fine for both integrated/GNU as. But if use it as it is, the ODR violation detection would be broken, because globals would have different addresses. There is a long discussion about this in GCC bugzilla (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=63888) and I don't know what solution is preferable here. Yury suggested to use names of globals instead of their addresses to detect ODR violation, is it an option here?

**Comment by kcc** [comment]
Dunno, maybe. 
At the very least it will be slower. Do you want to make a prototype and prove otherwise? 

Frankly, I am not motivated to invest time into this because in llvm everything works for the most common case (everything is instrumented) and in GCC the ODR checking is currently disabled. (Right?)

**Comment by ygribov2** [comment]
I'd say that sanitized is an important usecase as well and silently dropping support for it (especially in such an ugly way as in this bug) is a huge step back. How about we pass both global and local addresses to asan_register_global? Local address can then be used for poisoning and global for odr.

**Comment by kcc** [comment]
>> sanitized is an important usecase
Typo? 

>>  How about we pass both global and local addresses to asan_register_global?
That'll be another ABI break. Not a huge deal, but still.. 
I am still not sure how this is going to look. Want to prototype?

**Comment by ygribov2** [comment]
Yeah, I meant sanitized library.

I'm not sure why you care about ABI breaks but we could keep existing function and just add a new "extended" version to preserve compatibility. Max?

**Comment by chefmax** [comment]
So, we would just compare names to detect ODR violation? I think this can work (although we can still have ODR violation FP for C and others, but this is a different story). I can prepare a prototype and post it here.

**Comment by ygribov2** [comment]
No,  we'll still use global addresses for odr but locals for poisoning. That'll require a change to Asan ABI (register_global function).

**Comment by kcc** [comment]
>> I'm not sure why you care about ABI breaks
I don't care too much, I just don't want to do it w/o a good reason. 
>> I can prepare a prototype and post it here.
Not here, at llvm-commits, please

**Comment by ygribov** [comment]
LLVM's current approach to poisoning globals may also cause another tricky bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=68016 (yes, we've seen it in production).

**Comment by chefmax** [comment]
One another approach here:

1) Use private aliases for globals to be instrumented.
2) Pass an address to this alias to runtime to avoid poisoning of innocent globals in non-sanitized modules.
3) To preserve ODR violation checking, we introduce another global symbol (say, _asan_gen_XXX) and use it as indicator in runtime (say, 0 and 1).

Cons:
- The size of .bss section will be increased, we have doubled number of global symbols.
- ASan won't detect global buffer overflow in code with copy relocations due to using private aliases (see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=68016).
- ABI change.

Proc:
- This way, mixing sanitized code with non-sanitzed would be safe.

I have a prototype for this and if there aren't strong objections here, I'll post it at llvm-commits shortly.

**Comment by ygribov-samsung** [comment]
> The size of .bss section will be increased, we have doubled number of global symbols.

On the other hand, new globals would be 1-byte bools.

> ASan won't detect global buffer overflow in code with copy relocations due to using private aliases

Why? You could still poison copy-relocated global if it's __asan_gen_XXX is present.

**Comment by chefmax** [comment]
>> ASan won't detect global buffer overflow in code with copy relocations due to using private aliases
> Why? You could still poison copy-relocated global if it's __asan_gen_XXX is present.

Sorry, I've meant non-sanitized binary and sanitized library use case. For sanitized binary we indeed can poison copy-relocated global.

**Comment by ygribov-samsung** [comment]
> Sorry, I've meant non-sanitized binary and sanitized library use case.

Well, this case currently segfaults so I wouldn't treat this as a regression)

**Comment by chefmax** [comment]
This should be fixed on trunk now, can we close this?

**Comment by ygribov-samsung** [comment]
Another option is waiting until this is enabled by default.

**Comment by morehouse** [comment]
@kcc:  Do we want to enable `-asan-use-private-alias` by default?

**Comment by rnk** [comment]
I think this is the same as https://bugs.llvm.org/show_bug.cgi?id=37545

**Comment by MaskRay** [comment]
This issue can be closed now that -fsanitize-address-use-odr-indicator is the default for non-Windows platforms.
